### DISCUSSION TITLE
Getting the real IP address of a client (x-forwarded-for)

### DISCUSSION THREAD

**Discussion by teSill** [jchristn/WatsonTcp#246] [Q&A] [answered]
My server runs on CapRover using Docker and Nginx, and Nginx seems to be causing the client IP in IpPort to always be the server's IP (10.0.0.2 in this case). After some googling, it seems that CapRover's Nginx configuration allows the real IP to be passed in headers (https://github.com/caprover/caprover/issues/1077), but I'm not sure how to access it in WatsonTcp. The ConnectionEventArgs class only contains the IpPort string and seemingly no other data to work with.

**Comment by jchristn** [comment, marked as answer]
Hi, WatsonTcp only sees the IP of the TCP client to which it is peered, in your case, nginx.  I'm not sure how nginx can pass the actual source IP since there is no concept of an HTTP header (e.g. x-forwarded-for).

**Reply by teSill** [reply]
Hmm, that's tricky. Any way at all you can think of to identify specific clients? At least now with the new Guid approach instead of IpPort, there's no longer a risk of duplicate IpPorts but I'd still like to be able to determine if multiple connections are coming from the same address. In my case, I'm attempting to prevent players from logging onto the server with more than 2 accounts simultaneously.

**Reply by jchristn** [reply]
You would have to carry some identifying information within the messages you exchange or within the meta-data. I don't know that there's any other way since this is a TCP level library and not an HTTP level library.  The message metadata is essentially an equivalent to HTTP headers I would recommend, including some client unique identifier for their account as message metadata and check it server side.

**Reply by teSill** [reply]
Gotcha. Thank you for the quick replies!
